vision/standard_object/odlc_contour_detection.py

```diff
@@ -0,0 +1,136 @@
+def fetch_shape_contours(
+    image: Image, draw_contours: bool = False, resulting_file_name: str = ""
+) -> list[np.ndarray[np.dtype[np.uint8], np.dtype[np.uint8]]]:
+    """
+    Detects the boundaries of potential shapes based on a pixel or region's
+    brightness and saturation, finding only the darkest, brightest, and
+    most saturated portions of an image, and filters out shapes that are
+    guaranteed not to be the shapes we're expecting to see.
+
+    Parameters
+    ----------
+    filename : str
+        the name of the image to look for contours within
+    draw_contours : bool
+        True if we want to draw the contours and output the result
+
+    Returns
+    -------
+    contours : list[numpy.ndarray]
+        a list of all contours that could potentially be shapes we expect
+        contour : numpy.ndarray
+            an array of all points that make up the contour
+    """
+
+    hls_img = cv2.cvtColor(image, cv2.COLOR_RGB2HLS)  # converts image to HLS color format
+
+    # Blur is added to the image to reduce noise and to make the image/grass more uniform
+    new_img = cv2.GaussianBlur(hls_img, KERNEL_SIZE, sigmaX=0, sigmaY=0)  # blurs HLS image
+
+    img_brightness: ScImage
+    img_saturation: ScImage
+
+    img_brightness = np.array(new_img[:, :, 1])  # reads lightness of image as 2D array
+    img_saturation = np.array(new_img[:, :, 2])  # reads saturation of image as 2D array
+
+    # slightly blends blurred and unblurred images of same type, prioritizing blurred images
+    img_brightness = cv2.addWeighted(
+        hls_img[:, :, 1], NORMAL_IMG_WEIGHT, img_brightness, BLUR_IMG_WEIGHT, 0
+    )
+    img_saturation = cv2.addWeighted(
+        hls_img[:, :, 2], NORMAL_IMG_WEIGHT, img_saturation, BLUR_IMG_WEIGHT, 0
+    )
+
+    avg_brt: float = float(np.average(img_brightness))  # gets average brightness
+
+    white_thresh: np.ndarray[np.dtype[np.uint8], np.dtype[np.uint8]]
+    black_thresh: np.ndarray[np.dtype[np.uint8], np.dtype[np.uint8]]
+    saturation_thresh: np.ndarray[np.dtype[np.uint8], np.dtype[np.uint8]]
+
+    # Use higher saturation value if image is bright, lower if dark
+    saturation_value: int = (
+        MIN_SATURATION_VALUE if avg_brt > BRIGHTNESS_THRESH else MIN_DARK_SATURATION_VALUE
+    )
+
+    # White threshold is used to find the brightest parts of the image
+    # Black threshold is used to find the darkest parts of the image
+    # Saturation threshold is used to find the most saturated parts of the image
+    _, white_thresh = cv2.threshold(img_brightness, MIN_WHITE_VALUE, 255, cv2.THRESH_BINARY)
+    _, black_thresh = cv2.threshold(img_brightness, MAX_BLACK_VALUE, 255, cv2.THRESH_BINARY_INV)
+    _, saturation_thresh = cv2.threshold(img_saturation, saturation_value, 255, cv2.THRESH_BINARY)
+
+    # expands each threshold by 3 pixels in the x and y direction
+    # to merge those in close proximity to one another
+    white_thresh = cv2.dilate(white_thresh, np.ones([3, 3], np.uint8))
```

Comment:
make new constant `DILATION_KERNEL: NDArray[Shape["3, 3"], UInt8] = np.ones([3, 3], np.uint8)` and use it in all three of these